### PR TITLE
Random generation of NINOs

### PR DESCRIPTION
**Pull request by andrewgee**
NINO regex was also updated to require suffix of A-D.

### CODE REVIEW

File: project/HmrcBuild.scala [hunk 1]
```diff
@@ -50,10 +51,11 @@ object Dependencies {
   sealed abstract class Test(scope: String) {
     val scalaTest = "org.scalatest" %% "scalatest" % "2.2.4" % scope
     val pegdown = "org.pegdown" % "pegdown" % "1.5.0" % scope
+    val scalaCheck = "org.scalacheck" %% "scalacheck" % "1.12.5" % scope
```
Comment (SteveSmithTech) [inline project/HmrcBuild.scala:54]: Is scalaCheck needed as an import? It is not transitive from scalaTest?

Comment (andrewgee) [inline project/HmrcBuild.scala:54]: We thought that as well, but it seems to be needed for Prop.forAll

File: src/main/scala/uk/gov/hmrc/domain/Random.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.domain
+
+class Random(random: scala.util.Random = new scala.util.Random) {
+  def this(seed: Int) = this(new scala.util.Random(seed))
+
+  private val invalidFirstPrefixCharacters = List("D", "F", "I", "Q", "U", "V")
```
Comment (SteveSmithTech) [inline src/main/scala/uk/gov/hmrc/domain/Random.scala]: It feels a bit wrong to have Nino#invalidPrefixes and Random#invalidFirstPrefixCharacters, can they be unified somehow?

File: src/main/scala/uk/gov/hmrc/domain/Nino.scala [hunk 3]
```diff
@@ -33,8 +33,8 @@ object Nino {
   implicit val ninoWrite: Writes[Nino] = new SimpleObjectWrites[Nino](_.value)
   implicit val ninoRead: Reads[Nino] = new SimpleObjectReads[Nino]("nino", Nino.apply)
 
-  private val validNinoFormat = "[[A-Z]&&[^DFIQUV]][[A-Z]&&[^DFIQUVO]] ?\\d{2} ?\\d{2} ?\\d{2} ?[A-Z]{1}"
-  private val invalidPrefixes = List("BG", "GB", "NK", "KN", "TN", "NT", "ZZ")
+  private val validNinoFormat = "[[A-Z]&&[^DFIQUV]][[A-Z]&&[^DFIQUVO]] ?\\d{2} ?\\d{2} ?\\d{2} ?[A-D]{1}"
+  val invalidPrefixes = List("BG", "GB", "NK", "KN", "TN", "NT", "ZZ")
```
Comment (charleskubicek) [inline src/main/scala/uk/gov/hmrc/domain/Nino.scala]: I don't think this should be publicly available outside of this library as they are implementation details. Can we make it package protected or something similar?

Comment (andrewgee) [inline src/main/scala/uk/gov/hmrc/domain/Nino.scala]: Yes, this seems good, we'll change that.

File: src/main/scala/uk/gov/hmrc/domain/Random.scala (hunk 2 above)
Comment (andrewgee) [inline src/main/scala/uk/gov/hmrc/domain/Random.scala]: Yes, we'll move the invalid characters into Nino.

File: src/main/scala/uk/gov/hmrc/domain/Nino.scala [hunk 4]
```diff
@@ -33,8 +33,8 @@ object Nino {
   implicit val ninoWrite: Writes[Nino] = new SimpleObjectWrites[Nino](_.value)
   implicit val ninoRead: Reads[Nino] = new SimpleObjectReads[Nino]("nino", Nino.apply)
 
-  private val validNinoFormat = "[[A-Z]&&[^DFIQUV]][[A-Z]&&[^DFIQUVO]] ?\\d{2} ?\\d{2} ?\\d{2} ?[A-Z]{1}"
-  private val invalidPrefixes = List("BG", "GB", "NK", "KN", "TN", "NT", "ZZ")
+  private val validNinoFormat = "[[A-Z]&&[^DFIQUV]][[A-Z]&&[^DFIQUVO]] ?\\d{2} ?\\d{2} ?\\d{2} ?[A-D]{1}"
```
Comment (SteveSmithTech) [inline src/main/scala/uk/gov/hmrc/domain/Nino.scala]: Are we intentionally changing the Nino format here? Why?

Comment (andrewgee) [inline src/main/scala/uk/gov/hmrc/domain/Nino.scala]: This is because the regex was incorrect. Only suffixes can exist A-D.

Comment (SteveSmithTech) [inline src/main/scala/uk/gov/hmrc/domain/Nino.scala]: OK next time let's make that a separate PR

File: src/main/scala/uk/gov/hmrc/domain/Random.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.domain
+
+class Random(random: scala.util.Random = new scala.util.Random) {
```
Comment (howyp) [inline src/main/scala/uk/gov/hmrc/domain/Random.scala]: I think this class being called the same as the class of one of it's constructor arguments is a bit confusing and will also make it awkward to import for a client?

Perhaps we could call it `Generator`?

I'm thinking:

``` scala
trait Generate[T] {
  def random: scala.util.Random
  def generateNext: T
}

object Nino extends Generate[Nino] {
...
}
```

There might be a case for having `Generator[Nino]` available as a separate trait as well.

File: src/main/scala/uk/gov/hmrc/domain/Random.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+  private def generatePrefix: String = {
+    val prefix = generatePrefixCharacter(Nino.invalidFirstPrefixCharacters) +
+      generatePrefixCharacter(Nino.invalidSecordPrefixCharacters)
+    if (Nino.invalidPrefixes.contains(prefix)) generatePrefix
+    else prefix
+  }
+
+  def nextNino: Nino = {
+    val prefix = generatePrefix
```
Comment (SteveSmithTech) [inline src/main/scala/uk/gov/hmrc/domain/Random.scala]: I think the validation logic should be pushed inside Nino if possible. Then Random#nextNino could generate a prefix, generate a suffix, check if valid, then invoke random#nextInt to complete the NINO. That would simplify Random and allow it to support other tax identifiers in the near future

File: src/main/scala/uk/gov/hmrc/domain/Random.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  private def generatePrefixCharacter(invalidCharacters: List[String]): String = {
+    val character = generateCharacterByRange('A', 'Z')
+    if (invalidCharacters.contains(character)) generatePrefixCharacter(invalidCharacters)
+    else character
+  }
+
+  private def generatePrefix: String = {
+    val prefix = generatePrefixCharacter(Nino.invalidFirstPrefixCharacters) +
+      generatePrefixCharacter(Nino.invalidSecordPrefixCharacters)
+    if (Nino.invalidPrefixes.contains(prefix)) generatePrefix
```
Comment (charleskubicek) [inline src/main/scala/uk/gov/hmrc/domain/Random.scala]: Why do you generate a prefix then check if it's valid? Why not build a list of all valid characters at the beginning then choose random items from that list ?

File: src/test/scala/uk/gov/hmrc/domain/RandomSpec.scala [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.domain
+
+import org.scalacheck.Prop
+import org.scalatest.WordSpec
+import org.scalatest.prop.Checkers
+
+class RandomSpec extends WordSpec with Checkers {
```
Comment (SteveSmithTech) [inline src/test/scala/uk/gov/hmrc/domain/RandomSpec.scala]: This should be renamed as GeneratorSpec
